query-engine/prisma-models/src/field_selection.rs

```diff
@@ -0,0 +1,319 @@
+    pub fn is_superset_of(&self, other: &Self) -> bool {
+        other.selections.iter().all(|selection| match selection {
+            SelectedField::Scalar(sf) => self.contains(&sf.name),
+            SelectedField::Composite(other_cs) => self
+                .get(&other_cs.field.name)
+                .and_then(|selection| selection.as_composite())
+                .map(|cs| cs.is_superset_of(other_cs))
+                .unwrap_or(false),
+        })
+    }
+
+    pub fn selections(&self) -> impl Iterator<Item = &SelectedField> + '_ {
+        self.selections.iter()
+    }
+
+    /// Returns all Prisma (e.g. schema model field) names of contained fields.
+    /// Does _not_ recurse into composite selections and only iterates top level fields.
+    pub fn prisma_names(&self) -> impl Iterator<Item = String> + '_ {
+        self.selections.iter().map(|f| f.prisma_name().to_owned())
+    }
+
+    /// Returns all database (e.g. column or document field) names of contained fields.
+    /// Does _not_ recurse into composite selections and only iterates level fields.
+    pub fn db_names(&self) -> impl Iterator<Item = String> + '_ {
+        self.selections.iter().map(|f| f.db_name().to_owned())
+    }
+
+    /// Checked if a field of prisma name `name` is present in this `FieldSelection`.
+    pub fn contains(&self, name: &str) -> bool {
+        self.get(name).is_some()
+    }
+
+    pub fn get(&self, name: &str) -> Option<&SelectedField> {
+        self.selections.iter().find(|selection| selection.prisma_name() == name)
+    }
+
+    pub fn as_fields(&self) -> Vec<Field> {
+        self.selections()
+            .map(|selection| match selection {
+                SelectedField::Scalar(sf) => sf.clone().into(),
+                SelectedField::Composite(cf) => cf.field.clone().into(),
+            })
+            .collect()
+    }
+
+    /// Checks if `self` only contains scalar field selections and if so, returns them all in a list.
+    /// If any other selection is contained, returns `None`.
+    pub fn as_scalar_fields(&self) -> Option<Vec<ScalarFieldRef>> {
+        let scalar_fields = self
+            .selections()
+            .filter_map(|selection| match selection {
+                SelectedField::Scalar(sf) => Some(sf.clone()),
+                SelectedField::Composite(_) => None,
+            })
+            .collect_vec();
+
+        if scalar_fields.len() == self.selections.len() {
+            Some(scalar_fields)
+        } else {
+            None
+        }
+    }
+
+    /// Inserts this `FieldSelection`s selections into the given `SelectionResult`.
+    /// Assumes caller knows that the exchange can be done, but still errors if lengths mismatch.
+    /// Additionally performs a type coercion based on the source and destination field types.
+    /// Resistance is futile.
+    pub fn assimilate(&self, values: SelectionResult) -> crate::Result<SelectionResult> {
+        if self.selections.len() != values.len() {
```

Comment:
Very Minor, but you can avoid the else by returning the error. One let set of brackets makes it a little easier to reason about.